File: step1/leetcode/python/dp_lc53_maximum_subarray.py
#-*- coding=utf-8 -*-
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #idea:
        #f[n] = f[n-1] + n if n > 0, 选择n并加上
        #f[n] =          n if n < 0, 选择n并抛弃原来前面的数（因为是求连续的子序列）
        
        # 边界问题:
        # f[0] = 0
        
        f = len(nums)*[0]
        for i in range(0,len(nums)):
            if i == 0:
                f[i] = nums[i]
            A = f[i-1] + nums[i]
            B = nums[i]
            f[i] = max(A,B)
        return max(f)

        
def test_maxSubArray():
    input_array = [-2,1,-3,4,-1,2,1,-5,4]
    sln = Solution()
    ans = sln.maxSubArray(input_array)
    print(ans)

if __name__ == "__main__":
    test_maxSubArray()

File: step1/leetcode/cpp/dp_lc53_maximum_subarray.cpp
#include <iostream>
#include <vector>
#include <limits.h>
#include <algorithm>

using namespace std;
class Solution {
public:
    int maxSubArray2(vector<int>& nums) {
        int result = INT_MIN;
        vector<int> f(nums.size());
        int A,B;
        for(int i = 0; i< nums.size(); i++)
        {
            if (i == 0){
                f[i] = nums[i];
            }
            A = f[i-1] + nums[i];
            B = nums[i];
            f[i] = max(A,B);
        }
        return *max_element(f.begin(),f.end());
    }

    int maxSubArray(vector<int>& nums) {
        int result = INT_MIN;
        int f = 0;
        for(int i = 0; i< nums.size(); i++)
        {
            f = max(f + nums[i],nums[i]);
            result = max(f,result);
        }
        return result;
    }
};

void test_maxArray(){
    vector<int> input_array = {-2,1,-3,4,-1,2,1,-5,4};
    Solution* sln = new Solution();
    int result = sln->maxSubArray(input_array);
    cout << result << endl;
}


int main(){
    test_maxArray();
    return 0;
}

